File: api/src/main/java/net/blueshell/api/model/JobExecution.java
package net.blueshell.api.model;

import jakarta.persistence.*;
import lombok.Data;
import net.blueshell.api.base.BaseModel;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "job_executions")
@Data
@EntityListeners(AuditingEntityListener.class)
public class JobExecution implements BaseModel<UUID> {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "job_class", nullable = false)
    private String jobClass;

    @Column(name = "parameters", columnDefinition = "JSON")
    private String parameters;

    @Column(name = "status", nullable = false)
    @Enumerated(EnumType.STRING)
    private ExecutionStatus status;

    @Column(name = "started_at", nullable = false)
    private LocalDateTime startedAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    @Column(name = "execution_time_ms")
    private Long executionTimeMs;

    @Column(name = "retry_attempt", nullable = false)
    private int retryAttempt = 0;

    @Column(name = "error_message", columnDefinition = "TEXT")
    private String errorMessage;

    @Column(name = "stack_trace", columnDefinition = "TEXT")
    private String stackTrace;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    public enum ExecutionStatus {
        STARTED, COMPLETED, FAILED, RETRYING
    }
}

File: api/src/main/kotlin/net/blueshell/api/model/job/JobExecution.kt
package net.blueshell.api.model.job

import jakarta.persistence.*
import net.blueshell.api.model.base.AuditedVersionedEntity
import java.time.Instant

@Entity
@Table(
    name = "job_executions",
    indexes = [
        Index(name = "idx_job_executions_status", columnList = "status"),
        Index(name = "idx_job_executions_job_type", columnList = "job_type"),
        Index(name = "idx_job_executions_created_at", columnList = "created_at"),
    ]
)
class JobExecution(
    @Column(name = "job_type", nullable = false)
    var jobType: String = "",

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: JobExecutionStatus = JobExecutionStatus.QUEUED,

    @Lob
    @Column
    var payload: String? = null,

    @Lob
    @Column(name = "error_message")
    var errorMessage: String? = null,

    @Column(nullable = false)
    var attempts: Int = 0,

    @Column(name = "queued_at")
    var queuedAt: Instant? = null,

    @Column(name = "started_at")
    var startedAt: Instant? = null,

    @Column(name = "finished_at")
    var finishedAt: Instant? = null
) : AuditedVersionedEntity() {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
        protected set
}
